rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+  slug: 'my-discover-query', // Optional human-readable URL slug.
+  generator: 'DISCOVER_DEEP_LINKS',
+  state: {
+    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx',
+  },
+});
+```
+
+Both of the code snippets above would look the same and work on the server and
+browser, wherever the developer needs those.
+
+Finally, the Elastic Cloud would be able to use the redirect endpoint to
+navigate deeply within the Discover app. It could use the server-side HTTP
+endpoint:
+
+```
+POST /api/url/_redirect
+{
+  "generator": "DISCOVER_DEEP_LINKS",
+  "state": {
+    "indexPattern": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx"
+  }
+}
+```
+
+Or using the dedicated client-side "redirect" app:
+
+```
+GET /app/goto/_redirect/DISCOVER_DEEP_LINKS?state=(indexPattern: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx")
+```
+
+
+# Motivation
+
+Our motivation to improve the URL services comes from us intending to use them
+more, for example, for panel sharing to Slack or e-mail; and we believe that the
+current state of the URL services needs an upgrade.
+
+
+## Limitations of the Short URL Service
+
+We have identified the following limitations in the current implementation of
+the Short URL Service:
+
+1. It exposes only HTTP endpoint API.
+   1. __Will do:__ We will also expose a URL Service client through plugin
+      contract on the server and browser.
+1. It only has 3 HTTP endpoints, yet all three have different paths:
+   (1) `/short_url`, (2) `/shorten_url`; and (3) `/goto`.
+   1. __Will do:__ We will normalize the HTTP endpoints. We will use HTTP
+      method "verbs" like POST, instead of verbs in the url like "shorten_url".
+1. There is no migration system. If an application exposes this functionality,
+   every possible URL that might be generated should be supported forever. A
+   migration could be written inside the app itself, on page load, but this is a
+   risky path for URLs with many possibilities.
+   1. __Will do:__ Short URLs will be created using URL generators. We will use
+      migrations provided by the URL generators to migrate the stored parameters
+      in the short URL saved object.
+1. There is not much documentation for developers.
+   1. __Will do:__ The new service will have a much nicer API and docs.
+1. There is no way to delete short URLs once they are created.
+   1. __Will do:__ The new service will provide CRUD API to manage short URLs,
+      including deletion.
+1. Short URL service uses MD5 algorithm to hash long URLs. Security team
+   requested to stop using that algorithm.
+   1. __Will do:__ The new URL Service will not use MD5 algorithm.
+1. Short URLs store only the URL of the destination page. However, the
+   destination page might have other state which affects the display of the page
+   but is not present in the URL. Once the short URL is used to navigate to that
+   page, any state that is kept only in memory is lost.
+   1. __Could do:__ We will not support storing additional, non-URL state in
```

Comment:
if url generators would not be returning urls (so, they probably shouldn't be called url generators as mentioned in one of my ther comments) we can avoid this.

Reply:
Yes, this will be in the initial phase. I've first marked it as "Could do", but then actually expanded the API and talking with @ppisljar it will be in the initial phase.